<a href="https://colab.research.google.com/github/DavidSenseman/BIO1173/blob/main/Multiclass_Classification_Spring25_Assigment_01_LiverDisease.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---------------------------
**COPYRIGHT NOTICE:** This Jupyterlab Notebook is a Derivative work of [Jeff Heaton](https://github.com/jeffheaton) licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except in compliance with the License. You may obtain a copy of the License at

> [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

------------------------

# **BIO 1173: Intro Computational Biology**

**Assignment 1:  Neural Networks for Analysis of Tabular Data**

* Instructor: [David Senseman](mailto:David.Senseman@utsa.edu), [Department of Integrative Biology](https://sciences.utsa.edu/integrative-biology/), [UTSA](https://www.utsa.edu/)



# **READ CAREFULLY**

The **_first_**  digit in your myUTSA ID (e.g. "abc123") will determine which dataset you are to analyze for this assignment and which type of neural network (i.e. classification or regression) you will need to construct. For example, if your myUTSA ID was **vue682**, then your first digit is the number `6`.

**---WARNING------WARNING------WARNING------WARNING------WARNING------WARNING---**

You are **not** free to choose any dataset for this assignment. If analyze the wrong dataset, your assignment will **NOT BE GRADED**. If you are uncertain which dataset you should be working on, contact your Instructor for help. Remember, your score in this assignment will have a large impact on your course grade so please be careful.


| First Digit myUTSA ID    | Dataset to Analyze      | Neural Network Type
--------------------------|-------------------------|-----------------
0                         | Hepatitis               | Binary Classification
1                         | Coimbra Breast Cancer   | Binary Classification
2                         | Parkinson Speech        | Binary Classification
3                         | Indian Liver            | Binary Classification
4                         | Thyroid Replacement     | Multiclass Classification
5                         | Wine Quality            | Multiclass Classification
6                         | Liver Disease           | Multiclass Classification
7                         | Bone Marrow Transplant  | Regression
8                         | German Breast Cancer    | Regression
9                         | Diabetes Progression    | Regression

#### **NOTE: You can only use this Colab notebook if the first digit of your _myUTSA_ ID  is between `4` and `6`.**

# **The Purpose of Assignments**

In this course, **_Assignments_** are designed to help me (and you) assess your ability to transfer knowledge gained in completing class coding exercises to solving more realistic problems.

Assignments play a pivotal role in reinforcing your learning, as they require you to apply theoretical concepts to practical scenarios. This helps solidify your understanding and enhances your problem-solving skills. By tackling these assignments independently, you develop critical thinking and the ability to synthesize information from various sources. Moreover, assignments encourage you to explore topics more deeply, fostering intellectual curiosity and promoting a deeper engagement with the subject matter. Ultimately, these assignments are not just a measure of your learning, but a means to equip you with the skills needed for real-world applications and future challenges.

## **MAKE A COPY OF THIS NOTBOOK!**

For your assignment to be graded, you **must** make a copy of this Colab notebook in your GDrive and you this copy as your worksheet.

## Google CoLab Instructions

You MUST run the following code cell to get credit for this class lesson. By running this code cell, you will map your GDrive to /content/drive and print out your Google GMAIL address. Your Instructor will use your GMAIL address to verify the author of this class lesson.

In [1]:
# YOU MUST RUN THIS CELL FIRST

try:
    from google.colab import drive
    drive.mount('/content/drive', force_remount=True)
    from google.colab import auth
    auth.authenticate_user()
    COLAB = True
    print("Note: using Google CoLab")
    import requests
    gcloud_token = !gcloud auth print-access-token
    gcloud_tokeninfo = requests.get('https://www.googleapis.com/oauth2/v3/tokeninfo?access_token=' + gcloud_token[0]).json()
    print(gcloud_tokeninfo['email'])
except:
    print("**WARNING**: Your GMAIL address was **not** printed in the output below.")
    print("**WARNING**: You will NOT receive credit for this assignment.")
    COLAB = False

Mounted at /content/drive
Note: using Google CoLab
david.senseman@gmail.com


Your GMAIL address **must** appear in the output in order for your work to be graded.

# **Assigment 1: Multiclass Classification**

**Assignment_01** is specifically designed to assess your ability to write the Python/Tensorflow/Keras code necessary to build neural networks that can analyze tabular data stored in a Pandas DataFrame. These analyzes include: (1) binary classification, (2) multiclass classification and (3) regression.

You will use this Colab notebook **only** if the first digit in your myUTSA ID is between `4` and `6`. If that is correct, you have been assigned to perform **multiclass classification**.

Unlike your class lessons, you will **not** be given examples that you can use to simply copy-and-paste code. Rather, you will be given a problem to solve and it will be up to you to use code snippets that you have been given previously to solve different aspects of this assignment. And unlike your class lessons, your will **not** be given the correct output. In other words, this assignment is basically how you would solve an actual biomedical problem.

# **Multiclass Classification by Neural Networks**
**Multiclass classification** of tabular data is a type of supervised learning task where the goal is to categorize observations into one of three or more possible classes based on their attributes.

#### **Multiclass Classification:**
Multiclass classification deals with problems where there are multiple possible outcomes. Examples include:
1. **Disease Diagnosis:**
   - What type of diabetes does the patient have? (Type 1/Type 2/Gestational)
   - What stage is this breast tumor in? (Stage 0/Stage 1/Stage 2/Stage 3/Stage 4)
   - What type of hypertension does the patient have? (Primary/Secondary)
   - What variant of COVID-19 is the patient infected with? (Alpha/Beta/Delta/Omicron)
2. **Medical Outcomes:**
   - What is the prognosis for the patient after a heart attack? (Good/Fair/Poor)
   - What type of response will the patient have to a specific treatment? (Positive/Negative/Neutral)
   - What risk level does the patient have for developing heart disease? (Low/Medium/High)
3. **Medical Conditions and Symptoms:**
   - What type of depression is the patient experiencing? (Mild/Moderate/Severe)
   - What severity of sleep apnea does the patient have? (Mild/Moderate/Severe)
   - What is the patient's risk level for osteoporosis? (Low/Medium/High)
   - What genetic condition does the patient have a predisposition for? (Condition A/Condition B/Condition C)
4. **Medical Procedures:**
   - What type of surgery is recommended for this patient? (Surgery A/Surgery B/Surgery C)
   - What blood transfusion type does the patient need? (Type A/Type B/Type AB/Type O)
   - What category does the patient fall into for a particular clinical trial? (Group 1/Group 2/Group 3)

Here’s a step-by-step guide on how to perform multiclass classification using neural networks:
#### **Data Preparation:**
- **Collect Data:** Obtain a dataset with numerical features and a categorical target variable with multiple classes.
- **Clean Data:** Handle missing values, outliers, and erroneous entries.
- **Data Normalization:** Normalize your data (e.g. convert to Z-scores) to help the neural network learn more efficiently.
- **Data Pre-Processing:** Create X- and Y-feature vectors and encode the target variable into numerical format (e.g., one-hot encoding).
- **Split Data:** Divide your data into training and test sets.
#### **Neural Network Model**
- **Build the Neural Network Model:** Use TensorFlow and Keras to define the neural network architecture suitable for multiclass classification (e.g., using softmax activation in the output layer).
- **Train the Model:** Fit the model to your training data, using the validation set to monitor performance.
- **Evaluate the Model:** Assess the model’s performance on the test set using metrics like accuracy, precision, recall, F1-score, and Confusion Matrix:


## **Descriptions of Data Sets for Multiclass Classification**

This section describes the various datasets, information for downloading them, and what variable(s) your network should predict. Remember, you do **not** earn and credit if you analyze the wrong dataset. Pay particular attention to the **output** variable for Multiclass Classification because this column will contain the `Y- values` for your assigned dataset. You will need to know the name of the output feature when you are constructing yor `X-` and `Y-feature vectors`.

## **Thyroid Replacement Dataset - 1st myUTSA Digit = 4**

#### **Filename:** `thyroid_replacement.csv`
#### **Output Variable (Y values):** `class`


### **Thyroid Replacement Dataset**

The Thyroid Replacement dataset contains patient records and clinical features used to predict thyroid hormone replacement therapy. The dataset includes demographic information and various thyroid-related clinical test results.

#### Features:
1. **age**: Age of the patient.
2. **sex**: Gender of the patient (male/female).
3. **t_med**: Use of thyroxine (Yes/No).
5. **ant-med**: Use of antithyroid medication (Yes/No).
6. **sick**: Whether the patient is sick (Yes/No).
7. **pregnant**: Whether the patient is pregnant (Yes/No).
8. **t_surgery**: Whether the patient had thyroid surgery (Yes/No).
9. **I131_treat**: Whether the patient had radioactive iodine treatment (Yes/No).
12. **lithium**: Whether the patient is on lithium treatment (Yes/No).
13. **goiter**: Presence of goiter (Yes/No).
14. **turmor**: Presence of tumor (Yes/No).
15. **hypopituitary**: Whether the patient is hypopituitary (Yes/No).
16. **psych**: Whether the patient has psychiatric disorders (Yes/No).
17. **TSH**: Thyroid Stimulating Hormone level (mU/L).
18. **T3**: Triiodothyronine level (ng/dL).
19. **TT4**: Total Thyroxine level (μg/dL).
20. **FTI**: Free Thyroxine Index (unitless).
21. **TBG**: Free Thyroxine Index (unitless).


#### Output (Y variable):
- **class**: Indicates the thyroid hormone replacement therapy category (1: Class 1, 2: Class 2, 3: Class 3, 4: Class 4, 5: Class 5).


## **Wine Quality Dataset - 1st myUTSA Digit = 5**

#### **Filename:** `wine_quality.csv`
#### **Output Variable (Y values):** `quality`


### **Wine Quality Dataset**

The Wine Quality dataset contains physicochemical properties of wines, which are used to predict the quality of the wine. The dataset includes two sets of wines: red and white vinho verde wines from the north of Portugal. Each wine sample is described by 11 features and a quality score.

#### Features:
1. **color**: Wine color, red or white.
2. **fixed acidity**: Fixed acids that remain in the wine during fermentation (e.g., tartaric acid).
3. **volatile acidity**: Acetic acid content, which can give wine an undesirable vinegar flavor.
4. **citric acid**: Provides freshness to wines and can contribute to flavor.
5. **residual sugar**: The amount of sugar remaining after fermentation.
6. **chlorides**: The amount of salt in the wine.
7. **free sulfur dioxide**: SO₂ that is not bound to other molecules and is available to act as an antimicrobial agent.
8. **total sulfur dioxide**: The sum of free and bound forms of SO₂.
9. **density**: The density of the wine, closely related to alcohol and sugar content.
10. **pH**: Measures the acidity or alkalinity of wine.
11. **sulphates**: A wine preservative and antioxidant.
12. **alcohol**: Alcohol content of the wine.

#### Output (Y variable):
- **quality**: Quality score of the wine, likely on a scale from 0 to 10, based on sensory data from wine experts.


## **Liver Disease Dataset - 1st myUTSA Digit = 6**

#### **Filename:** `liver_disease.csv`
#### **Output Variable (Y values):** `drinks`

### **Liver Disease Dataset**

The Liver Disease dataset contains patient records and clinical features used to predict the presence of liver disease. The dataset includes various biochemical markers and a classification label indicating liver disease status.

#### Features:
1. **class**: Indicates the presence of liver disease (`Diseased`, `No_Disease`)
2. **mcv**: Mean Corpuscular Volume (fl)
3. **alkphos**: Alkaline Phosphatase levels (IU/L)
4. **sgpt**: Serum Glutamic-Pyruvic Transaminase (ALT) levels (IU/L)
5. **sgoit**: Serum Glutamic-Oxaloacetic Transaminase (AST) levels (IU/L)
6. **gammagt**: Gamma-Glutamyl Transferase levels (IU/L)

#### Output (Y variable):
- **drinks**: Number of alcoholic drinks consumed per day-


# **General Instructions**

To make the assignment more manageable, you will given a number of specific steps to perform. To help guide you in writing your code, you will be given a specific example in a particular class lesson that you can use for a reference. For example, in **Step 1: Download and Extract Data** you are given **REF: Class_01_6 (Example 1)**. That means Example 1 in Class_01_6 provides similar code that you could use to complete that step of this assignment.

### **Variable Names**

In writing your code for this assignment, you are free to give your variables any name that makes sense to you. This includes the name of the DataFrame that holds your data. When you `copy-and-paste` code from earlier Class assignments, you always have to edit the name of the DataFrame to match the name you select for this assignment in **Step 1**.

When it has been necessary to give an example that includes a DataFrame name, the DataFrame has been called `dataFrameDF`. You will need to edit the name `dataFrameDF` to match the actual name you have given to your DataFrame in **Step 1**.

### **Can I Use AI?**

You are free to use AI (e.g. Microsoft Co-Pilot) to help you complete your assignment---but you need to be very careful.

While AI can be very helpful in correcting coding errors, but it can also give you code that is totally incorrect for this assignment. A small number of students in previous classes have flunked their assignment by using AI code that did not generate the correct output. Useless you give the AI a well-constructed prompt, the answer you get back might lead you in the wrong direction.

If your aren't sure what you are doing, it's much, much safer to get help with any of your coding problems from your course instructor and/or course TA's.

### Define functions

The cell below creates several functions that are needed for this assignment. If you don't run this cell, you will receive errors later when you try to run some cells.

In [18]:
# Create functions for this lesson

%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.metrics import roc_curve, auc


def list_float_columns(dataframe):
    """
    Create a list of all columns in a DataFrame that contain float values.

    Parameters:
    dataframe (pd.DataFrame): The DataFrame to check.

    Returns:
    list: A list of column names that contain float values.
    """
    float_columns = [col for col in dataframe.columns if dataframe[col].dtype == 'float64']
    return float_columns

### **STEP 1: Download and Extract Data**

**REF: Class_01_6 (Example 1)**

As usual, a coding project starts with downloading a dataset. Since your dataset is in tabular form, you should use Pandas to read the datafile and store the information in a DataFrame. You are free to choose the name for all or your variables in this assignment, including the name of your DataFrame.

In the cell below, write the code to download your datafile from the course server and create a Pandas DataFrame to store your data. Use the function `display()` to show the data in 8 rows. For full credit, you need to show **all** of the columns in your DataFrame. You can do this by using the following code:

~~~text
# Set max columns and max rows
pd.set_option('display.max_columns', dataFrameDF.shape[1])
pd.set_option('display.max_rows', 8)
~~~

where `dataFrameDF` is the name of your DataFrame.


In [19]:
# Step 1: Download and Extract Data

import numpy as np
import pandas as pd

# Read file and create DataFrame
ldDF = pd.read_csv(
    "https://biologicslab.co/BIO1173/data/liver_disease.csv",
#    index_col=0,
    na_values=['NA','?'])

# Set max columns and max rows
pd.set_option('display.max_columns', ldDF.shape[1])
pd.set_option('display.max_rows', 8)

# Display DataFrame
display(ldDF)

class  mcv  alkphos  sgpt  sgoit  gammagt  drinks
0      Diseased   85     92.0    45     27       31       0
1    No_Disease   85     64.0    59     32       23       0
2    No_Disease   86     54.0    33     16       54       0
3    No_Disease   91     78.0    34     24       36       0
..          ...  ...      ...   ...    ...      ...     ...
341  No_Disease   96     69.0    53     43      203      12
342    Diseased   98     77.0    55     35       89      15
343    Diseased   91     68.0    27     26       14      16
344    Diseased   98     99.0    57     45       65      20

[345 rows x 7 columns]

If your code is correct you should see a table with a relatively large number of columns that may extend beyond the right edge of your notebook.

## **STEP 2: Describe DataFrame**

**REF: Class_01_6 (Example 3)**

The `df.describe()` command in Python is used with pandas DataFrames. It provides a summary of statistics for each column in the DataFrame. By default, it will return the count, mean, standard deviation, min, 25th percentile (Q1), median (50th percentile), 75th percentile (Q3), and max values for numerical columns. It can be a handy tool for getting a quick overview of your dataset!

Use the `df.describe()` command to summaries the data in your DataFrame. Make sure to replace the `df` with the actual name of your DataFrame.

Again use these commands to set the display options:

~~~text
# Set max columns and max rows
pd.set_option('display.max_columns', dataFrameDF.shape[1])
pd.set_option('display.max_rows', 8)
~~~


In [20]:
# Step 2: Describe DataFrame

import pandas as pd

# Set max columns and max rows
pd.set_option('display.max_columns', ldDF.shape[1])
pd.set_option('display.max_rows', 8)

# Describe() method
ldDF.describe()

mcv     alkphos        sgpt       sgoit     gammagt      drinks
count  345.000000  340.000000  345.000000  345.000000  345.000000  345.000000
mean    90.159420   70.008824   30.405797   24.643478   38.284058    3.611594
std      4.448096   18.410829   19.512309   10.064494   39.254616    3.204264
min     65.000000   23.000000    4.000000    5.000000    5.000000    0.000000
25%     87.000000   57.000000   19.000000   19.000000   15.000000    1.000000
50%     90.000000   67.000000   26.000000   23.000000   25.000000    3.000000
75%     93.000000   80.000000   34.000000   27.000000   46.000000    6.000000
max    103.000000  138.000000  155.000000   82.000000  297.000000   20.000000

If your code is correct you should see a table with a relatively large number of columns that may extend beyond the right edge of your notebook and 8 rows countaining the summary statistics for each column.

## **STEP 3: Find Missing Values**

**REF: Class_01_6 (Example 4)**

In **Biostatistics**, finding and replacing missing values is crucial for several reasons:

1. **Preserving Data Integrity**: Missing values can distort the analysis, leading to biased results. By addressing missing values, you ensure that your conclusions are based on complete and accurate data.

2. **Statistical Validity**: Many statistical tests and models require complete data. Missing values can reduce the statistical power of these tests, making it difficult to detect real effects.

3. **Avoiding Data Loss**: Simply discarding rows or columns with missing values can lead to a significant loss of data, especially if the dataset is already small. Imputing missing values helps retain as much data as possible.

4. **Model Accuracy**: Machine learning models can be sensitive to missing data. Handling missing values appropriately can improve the performance and accuracy of predictive models.

5. **Consistency**: Different columns in a dataset may have varying levels of missing data. Addressing these inconsistencies helps in creating a more uniform dataset, which is easier to analyze and interpret.

The `df.isnull()` command in pandas is used to detect missing values in a DataFrame. It returns a DataFrame of the same shape as the original, but with Boolean values: `True` where the value is missing (`NaN`) and `False` where the value is not missing.

In the cell below, use the command `df.isnull()` to find and print out any missing values in your DataFrame.

To make sure you see all of the values, use this code to set your display output:

~~~text
# Set max columns and max rows
pd.set_option('display.max_columns', dataFrameDF.shape[1])
pd.set_option('display.max_rows', dataFrameDF.shape[0])

~~~

In [21]:
# Step 3: Find Missing Values

import pandas as pd

# Find the locations of missing data
missing_locations = ldDF.isnull().any()

# Set max columns and max rows
pd.set_option('display.max_columns', ldDF.shape[1])
pd.set_option('display.max_rows', ldDF.shape[0])

# Display the locations of missing data
print(missing_locations)

class      False
mcv        False
alkphos     True
sgpt       False
sgoit      False
gammagt    False
drinks     False
dtype: bool


Inspect the your output and see if one (or more) columns have the word `True` after them. These columns contain one (or more) missing values. Make note of the column name since your will need to handle these missing values in the next step.  

## **STEP 4: Replace Missing Values**

**REF: Class_01_6 (Example 6)**

One common strategy for replacing missing values is to use the `median` of the column to replace the missing value. The `median` is used instead of the column `mean` since the `median` is a robust measure of central tendency and is less affected by outliers compared to the `mean`.

To do this:
1. Calculate the median of the column: Use the `median()` function to find the median value of the column.

2. Replace the missing values: Use the `fillna()` function to replace the missing values with the median.

If your DataFrame had one (or more) columns with missing values, you will need write the Python code to replace these missing values with the column's `median` value in the cell below.

After you have replaced the missing values, again use the `df.isnull()` command to print out the names of columns with missing values to make sure all of this missing values have been taken care of. (Just use the same code you wrote in **Step 3**.)




In [22]:
# Step 4: Replace Missing Values

import pandas as pd

# Find the median of the column Insulin
alkphos_med = ldDF['alkphos'].median()

# Print out the median value
print(f"The median value = {alkphos_med} for alkphos.")
print(f"Replacing missing values with {alkphos_med}.")

# Use fillna method
ldDF['alkphos'] = ldDF['alkphos'].fillna(alkphos_med)

# Find the locations of missing data
print("\nLooking for missing values...")  # The \n means print a newline
missing_locations = ldDF.isnull().any()

# Display the locations of missing data
print(missing_locations)

The median value = 67.0 for alkphos.
Replacing missing values with 67.0.

Looking for missing values...
class      False
mcv        False
alkphos    False
sgpt       False
sgoit      False
gammagt    False
drinks     False
dtype: bool


If your code is correct, you should see the same output as in **Step 3**, but this time all of the column names should be followed by the word `False`.

## **STEP 5: Display Non-numeric Categories**

**REF: Class_02_2 (Example 3 - Step 1)**

When building neural networks it is especially important to know which columns contain non-numeric data ("strings").

In the cell below, write the code to print out a list of columns in your DataFrame that contain non-numeric data.

In [23]:
# Step 5: Display non-numeric categories

import pandas as pd

# Select columns
non_numerical_columns = ldDF.select_dtypes(exclude='number').columns

# Print result
print(*non_numerical_columns)

class


## **STEP 6: Print Names in a Categorical Column**

**REF: Class_02_2 (Example 3 - Step 3)**

In the cell below, write the code to print out a list of categorical values (strings) that are in the non-numeric column found in **Step 5**.




In [24]:
# Step 6: Print names in a categorical column

import pandas as pd

# Generate a list with only unique values
numCat = list(ldDF['class'].unique())

# Print out the results
print(f'Number of categories: {len(numCat)}')
print(f'String names: {numCat}')

Number of categories: 2
String names: ['Diseased', 'No_Disease']


The output above gives the string names that you need to map to integers in next step.


## **STEP 7: Map Strings to Integer Values**

**REF: Class_02_2 (Example 3 - Step 2)**

In the cell below, write the code to map each string shown in the output above to either the integer `0` or `1`.

To make sure your mapping worked as intended, use the `display(df)` function to display your updated DataFrame.

In [25]:
# Step 7: Map Strings to Integer Values

# Set max columns and max rows
pd.set_option('display.max_columns', ldDF.shape[1])
pd.set_option('display.max_rows', 8)


# Define the mapping dictionary
mapping = {'Diseased': 0, 'No_Disease': 1}

# Check if all values to be mapped are present in the column
unique_values = ldDF['class'].unique()

# Find values in the column that are not in the mapping dictionary
missing_values = [value for value in unique_values if value not in mapping]

if missing_values:
    print(f"Error: The following values in the 'class' column are not in the mapping dictionary: {missing_values}")
    print(f"Error: Either your mapping is wrong or you have already converted the strings to integers")
else:
    # Map the 'Gender' column using the mapping dictionary
    ldDF['class'] = ldDF['class'].map(mapping)
    print("Liver disease data after mapping:")
    display(ldDF)

Liver disease data after mapping:


class  mcv  alkphos  sgpt  sgoit  gammagt  drinks
0        0   85     92.0    45     27       31       0
1        1   85     64.0    59     32       23       0
2        1   86     54.0    33     16       54       0
3        1   91     78.0    34     24       36       0
..     ...  ...      ...   ...    ...      ...     ...
341      1   96     69.0    53     43      203      12
342      0   98     77.0    55     35       89      15
343      0   91     68.0    27     26       14      16
344      0   98     99.0    57     45       65      20

[345 rows x 7 columns]

If your code is correct your output should be the same as **Step 1** except that the strings in the non-numeric column should now be integers.

## **STEP 8: Shuffle and Reindex your DataFrame**

**REF: Class_02_3 (Example 2)**

**Shuffling and reindexing** the data are important steps when building neural networks for several reasons:

1. **Preventing Overfitting**: When the data is in a specific order, the model might learn patterns that are a result of the order rather than the actual data. Shuffling helps to prevent the model from overfitting to these spurious patterns.

2. **Improving Generalization**: By shuffling the data, you ensure that the model is exposed to a wide variety of examples during training, which helps it to generalize better to new, unseen data.

3. **Reducing Bias**: Shuffling ensures that the distribution of data is more uniform across the training batches. This reduces the risk of introducing bias, especially if the data has some inherent order that might otherwise affect the model's performance.

4. **Enhancing Convergence**: Neural networks often train faster and more reliably when the data is shuffled. This is because the model updates its parameters in a more representative and balanced manner.

5. **Ensuring Robustness**: By reindexing the data, you avoid potential issues that could arise from relying on the original indices, which might have some underlying structure or grouping that could affect the training process.


In the cell below, write the code to shuffle and reindex your DataFrame. Then display your shuffled DataFrame using the following display settings:

~~~text
# Set max columns and max rows
pd.set_option('display.max_columns', dataFrameDF.shape[1])
pd.set_option('display.max_rows', 8)
~~~

In [26]:
# Step 8: Shuffle and Reindex your DataFrame

import pandas as pd
import numpy as np

# Set the random seed to 42
np.random.seed(42)

# Use random.permutation function for shuffling & reindexing
ldDF = ldDF.reindex(np.random.permutation(ldDF.index))

# Set max columns and max rows
pd.set_option('display.max_columns', ldDF.shape[1])
pd.set_option('display.max_rows', 8)

# Display the shuffled & reindexed DataFrame
display(ldDF)

class  mcv  alkphos  sgpt  sgoit  gammagt  drinks
194      0   97     62.0    17     13        5       1
157      1   92     87.0    57     25       44       6
225      1   85     58.0    18     24       16       1
208      0   89     82.0    33     32       18       1
..     ...  ...      ...   ...    ...      ...     ...
71       0   90     67.0    35     19       35       2
106      0   91     80.0    37     23       27       4
270      1   88     57.0     9     15       16       2
102      0   89     63.0    22     27       10       4

[345 rows x 7 columns]

If your code is correct, the output generated above by **Step 8** should look almost identical to the output you got after running **Step 7** except that this index numbers shown in the leftmore column should be in a random order.

## **STEP 9: Normalize Data to the Z-values**

**REF: Class_02_2 (Example 2)**

When building neural networks, it is important to normalize the data using Z-score normalization for several reasons:

1. **Standardization**: Z-score normalization standardizes the features to have a mean of 0 and a standard deviation of 1. This ensures that all features contribute equally to the model and prevents features with larger scales from dominating the learning process.

2. **Improved Convergence**: Neural networks often use gradient-based optimization algorithms. Normalized data can lead to more stable and faster convergence of these algorithms because the gradients are more balanced and less likely to explode or vanish.

3. **Enhanced Performance**: Normalized data often leads to better model performance. By bringing all features to a similar scale, the model can learn more effectively and produce more accurate predictions.

4. **Handling Outliers**: Z-score normalization reduces the impact of outliers by transforming the data to a standard scale. Outliers will have high positive or negative Z-scores, but their influence will be mitigated compared to unnormalized data.

5. **Consistency**: Normalization ensures that different features are on the same scale, which is particularly important when combining multiple datasets or using features with different units of measurement.

To make things easier, only normalize the `float` data in your dataset by converting these numbers into their Z-values.

In the cell below, use the function `list_float_columns()` to find the names of all of the columns in your DataFrame that contain float number. (NOTE: This is custom function that was created near the beginning of the assignment). Once you know the column names, use can use the `zscore()` function to normalize the values only in the columns containing floats. Print out the names of the columns containing floats, and the first five Z-scores in each of these columns.

In [27]:
# Step 9: Normalize Data to the Z-values

import pandas as pd
from scipy.stats import zscore

float_columns = list_float_columns(ldDF)
print(f"Columns with float values: {float_columns}")

for col in float_columns:
    ldDF[col] = zscore(ldDF[col])

# Print the first 5 values of each float column
for col in float_columns:
    print(f"First 5 values in column '{col}': {ldDF[col].head().tolist()}")

Columns with float values: ['alkphos']
First 5 values in column 'alkphos': [-0.4363647930615055, 0.9332299449863417, -0.655499951149161, 0.6593109973767722, 1.261932682117825]


**HINT:** If you see this error:

~~~text
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-8-9b058d8fbfea> in <cell line: 0>()
      4 from scipy.stats import zscore
      5
----> 6 float_columns = list_float_columns(dataFrameDF)
      7 print(f"Columns with float values: {float_columns}")
      8

NameError: name 'list_float_columns' is not defined
~~~~

it means you didn't run the cell called `Define functions` at the beginning of this assignment.


## **STEP 10: Pre-process Data for Neural Network Training**

**REF: Class_04_3 (Example 9)**

In the cell below, write the code to preprocess the data in your DataFrame to make it ready to feed into your neural network.

**NOTES: Please follow these directions carefully:**

1. Since you have already converted your float values into their Z-scores, you should **not** normalize any data during your pre-processing. In other words, converting Z-scores into Z-scores, a second time, is not a good thing.

2. Basically all you need to do is write the code to generate your `X-feature vector` and your `Y-feature vector`. Your `Y-values` will be in the column that is the output variable. The name of the output variable (`Y` values) for your particular dataset was specified in the dataset description at the start of this assignment.

3. When generating your `X-feature vector`, you should use _all_ of the columns in your DataFrame **EXCEPT** for the column containing the `Y-values`.

4. You do **not** need to use the code in the snippet: `Standardize ranges in numeric column`. You have already converted the numeric columns (in this case `floats`) to their Z-scores.

5. Since you will be building a **Multiclass Classification** neural network, you **must** one-hot encode the Y-values when generating your `Y-feature vector`.

6. Do **not** split your data into training and test set yet. You will do the split later.

7. When you are done, generating both your `X-` and `Y-` feature vectors, print out the first 4 values in each vector.

In [29]:
# Step 10 - Preprocess Data for Neural Network Training

import pandas as pd
from scipy.stats import zscore

# Generate column list for preprocessing
ldX_columns = ldDF.columns.drop('drinks')

# Generate X feature vector
ldX = ldDF[ldX_columns].values
ldX = np.asarray(ldX).astype(np.float32)

# One-Hot encode the target column and generate Y
dummies = pd.get_dummies(ldDF['drinks'], dtype=float) # Classification
ld_classes = dummies.columns
ldY = dummies.values
ldY = np.asarray(ldY).astype(np.float32)

# Print out X and Y
np.set_printoptions(suppress=True,precision=4)
print("The first 4 X-values are:")
print(ldX[0:4])
print("\nTheir corresponding Y-values are:")
print(ldY[0:4])

The first 4 X-values are:
[[ 0.     97.     -0.4364 17.     13.      5.    ]
 [ 1.     92.      0.9332 57.     25.     44.    ]
 [ 1.     85.     -0.6555 18.     24.     16.    ]
 [ 0.     89.      0.6593 33.     32.     18.    ]]

Their corresponding Y-values are:
[[0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]


## **STEP 11: Construct and Compile Neural Network**

**REF: Class_04_2 (Example 10)**

In the cell below, use the Keras/Tensorflow libraries to split your data into `test` and `train` splits, making the test size = 0.25, and set the random state to `42`.

Then construct and compile a binary classification neural network with 3 hidden layers but do **not** start your training in this step.

To construt your model, you can use the code chunk in Example 10 in the section that starts with:

~~~text
# Construct model-----------------------------------------------------
~~~
Since this neural network will perform multiclass classification, there should be same number of neurons in the output layer as their are Y classes.

After you construct it,  complie your model. To compile your model you can use the code chunk in Example 3 in the section that starts with:

~~~text
# Compile model------------------------------------------------------------------
~~~

Do **not** start training your model yet. This will be done in a separate step.

In [30]:
# Step 11: Construct and compile neural network

from keras.models import Sequential
from keras.layers import Dense, Input
from keras.callbacks import EarlyStopping
from keras.optimizers import Adam
from keras.metrics import Precision, Recall
from sklearn.model_selection import train_test_split
import numpy as np

# Split into train/test--------------------------------------------------------
ldX_train, ldX_test, ldY_train, ldY_test = train_test_split(
    ldX, ldY, test_size=0.25, random_state=10)


# Construct model---------------------------------------------------------------
ldModel = Sequential()
ldModel.add(Input(shape=(ldX.shape[1],)))
ldModel.add(Dense(100, activation='relu',
                kernel_initializer='random_normal'))  # Hidden 1
ldModel.add(Dense(50,activation='relu',
                   kernel_initializer='random_normal')) # Hidden 2
ldModel.add(Dense(25,activation='relu',
                   kernel_initializer='random_normal')) # Hidden 3
ldModel.add(Dense(ldY.shape[1],activation='softmax',
                kernel_initializer='random_normal')) # Output

# Compile model------------------------------------------------------------------
ldModel.compile(loss='categorical_crossentropy',
              optimizer=Adam(),
              metrics =['accuracy'])



If your code is correct, you should **not** see any output after running the previous cell.

## **STEP 12: Print Summary of Your Model**

**REF: Class_04_2**

The `model.summary()` command in deep learning frameworks like Keras and TensorFlow provides a detailed summary of the neural network model. This summary includes useful information about the model's architecture, including:

1. **Layer Names and Types:** The name and type (e.g., Dense, Conv2D, LSTM) of each layer in the model.

2. **Output Shape:** The shape of the output produced by each layer.

3. **Number of Parameters:** The total number of trainable and non-trainable parameters in each layer. This includes both the weights and biases.

4. **Model Parameters Summary:** A total count of all trainable and non-trainable parameters in the model.

In the cell below, use the `model.summary()` command to print out the information about your neural network.



In [31]:
# Step 12: Print Summary of Your Model

ldModel.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 100)                 │             700 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 50)                  │           5,050 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 25)                  │           1,275 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_3 (Dense)                      │ (None, 15)                  │             390 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 7,415 (28.96 KB)

 Trainable params: 7,415 (28.96 KB)

 Non-trainable params: 0 (0.00 B)

## **STEP 13: Create Early Stopping Monitor**

**REF: Class_04_2 (Example 9)**

An **Early Stopping Monitor** is a technique used during the training of neural networks to prevent overfitting and improve the model's generalization to new, unseen data. It works by monitoring the performance of the model on a validation dataset and stopping the training process when the performance starts to degrade.

#### Here’s how it works:

1. **Monitoring Performance**: Early stopping keeps track of a specific metric, such as validation loss or validation accuracy, during each epoch of training.

2. **Patience**: It has a parameter called "patience," which defines the number of epochs to wait for an improvement in the monitored metric before stopping the training. If the performance does not improve for a specified number of epochs, the training is stopped.

3. **Restore Best Weights**: In some implementations, early stopping can also restore the model weights to the state that resulted in the best performance on the validation set.

#### Benefits of early stopping include:

- **Preventing Overfitting**: By stopping training when the model starts to overfit the training data, early stopping helps maintain good generalization performance.
- **Saving Time and Resources**: It avoids unnecessary training epochs, saving computational resources and time.

In the cell below, write the code to create an Early Stopping Monitor that monitors `val_loss`. Set the parameter `patience` to `10`.



In [32]:
# Step 13: Create early stopping monitor

PATIENCE=10

ldMonitor = EarlyStopping(monitor='val_loss', min_delta=1e-3,
    patience=PATIENCE, verbose=1, mode='auto', restore_best_weights=True)

If your code is correct, you should not see any output.

## **STEP 14: Train the Model**

In the cell below, write the Python code to train the neural network that you constructed in **Step 11**. Set the number of epochs to `100`. Make sure the parameter `verbose` is set to `2` so that the output of each epoch is written out.

In [33]:
# Step 14: Train the Model

# Set variables
EPOCHS=100
VERBOSE=2

# Train model
ldModel.fit(ldX_train,ldY_train,validation_data=(ldX_test,ldY_test),
          callbacks=[ldMonitor],verbose=VERBOSE,epochs=EPOCHS)


Epoch 1/100
9/9 - 2s - 236ms/step - accuracy: 0.2674 - loss: 2.5828 - val_accuracy: 0.3103 - val_loss: 2.4915
Epoch 2/100
9/9 - 0s - 10ms/step - accuracy: 0.3798 - loss: 2.3194 - val_accuracy: 0.3103 - val_loss: 2.3283
Epoch 3/100
9/9 - 0s - 14ms/step - accuracy: 0.3798 - loss: 2.1670 - val_accuracy: 0.3103 - val_loss: 2.2333
Epoch 4/100
9/9 - 0s - 9ms/step - accuracy: 0.3798 - loss: 2.0977 - val_accuracy: 0.3103 - val_loss: 2.1663
Epoch 5/100
9/9 - 0s - 9ms/step - accuracy: 0.3798 - loss: 2.0747 - val_accuracy: 0.3103 - val_loss: 2.1280
Epoch 6/100
9/9 - 0s - 9ms/step - accuracy: 0.3798 - loss: 2.0373 - val_accuracy: 0.3103 - val_loss: 2.1301
Epoch 7/100
9/9 - 0s - 15ms/step - accuracy: 0.3798 - loss: 2.0116 - val_accuracy: 0.3103 - val_loss: 2.0978
Epoch 8/100
9/9 - 0s - 9ms/step - accuracy: 0.3837 - loss: 1.9961 - val_accuracy: 0.3218 - val_loss: 2.1007
Epoch 9/100
9/9 - 0s - 9ms/step - accuracy: 0.3837 - loss: 1.9863 - val_accuracy: 0.3218 - val_loss: 2.0918
Epoch 10/100
9/9 - 0s -

## **STEP 15: Convert Prediction Probabilites into Actual Prediction Values**

**REF: Class_04_2 (Example 11)**

In the cell below, write the code to convert the prediction probabilities from your neural network model into actual prediction probabilities using the `df.armax()` method. Print out the prediction probabilites for the first 6 samples, and then print out the prediction values for the same 6 samples.

In [34]:
# Step 15: Convert Prediction Probabilites into Actual Prediction Values

import tensorflow as tf

# Define the predict function outside of any loops to prevent retracing
@tf.function
def predict_probabilities(model, inputs):
    return model(inputs)

# Use model to predict probabilities for subjects in x_test
ldProb = predict_probabilities(ldModel, ldX_test)

# Print prediction probabilities (first 6 samples)
print(f"Prediction Probabilities (first 6):\n{ldProb[0:6]}")

# Use argmax to convert probabilities to prediction values
ldPred = tf.argmax(ldProb, axis=1)

# Print out prediction values (first 6 samples)
print(f"Prediction Values (first 6):\n{ldPred.numpy()[0:6]}")

Prediction Probabilities (first 6):
[[0.032  0.3709 0.1079 0.062  0.1737 0.0646 0.0888 0.0112 0.0408 0.0024
  0.0296 0.002  0.0053 0.0049 0.004 ]
 [0.0314 0.3474 0.1099 0.0605 0.1698 0.0672 0.0942 0.0133 0.0481 0.0035
  0.0337 0.0028 0.0067 0.0058 0.0055]
 [0.0327 0.3609 0.1083 0.0617 0.173  0.0646 0.0905 0.0122 0.0432 0.0028
  0.0318 0.0023 0.0059 0.0054 0.0047]
 [0.0352 0.2865 0.1126 0.0641 0.1537 0.0732 0.1038 0.02   0.0618 0.0073
  0.0439 0.0062 0.0116 0.0098 0.0103]
 [0.023  0.4596 0.0968 0.0525 0.1799 0.0542 0.0761 0.0058 0.0269 0.0007
  0.0181 0.0006 0.0022 0.0021 0.0015]
 [0.0351 0.3278 0.1133 0.0672 0.1636 0.0729 0.0944 0.0146 0.0487 0.004
  0.0346 0.0033 0.0076 0.0068 0.006 ]]
Prediction Values (first 6):
[1 1 1 1 1 1]


## **STEP 16: Compute the Percent Accuracy**

**REF: Class_04_2 (Example 11)**

In the cell below, compute the percent accuracy of your neural network model. Print out the first 6 values in your `Y_compare` variable as well as your percent accuracy score.

NOTE: **Step 16** uses the variable holding the actual prediction values generated in **Step 15**, so you need to run **Step 15** before your run this step.

In [35]:
# Step 16: Compute percent accuracy

import numpy as np
from sklearn import metrics

# Generate array containing actual class type
ldY_compare = np.argmax(ldY_test, axis=1)
print(f'First 6 values in Y_compare: {ldY_compare[0:6]}')

# Make sure ecgPred is also a NumPy array on the CPU
if not isinstance(ldPred, np.ndarray):
    ldPred = ldPred.numpy()

# Compute the percentage score
score = metrics.accuracy_score(ldY_compare, ldPred)

# Print out the score
print("Accuracy score: {}".format(score))

First 6 values in Y_compare: [1 1 1 0 1 6]
Accuracy score: 0.3103448275862069


## **Assignment Turn-in**

When you have completed and run all of the code cells, use the **File --> Print.. --> Save to PDF** to generate a PDF of your Colab notebook. Save your PDF as `Copy of Assignment_01.lastname.pdf` where _lastname_ is your last name, and upload the file to Canvas.

## **Poly-A Tail**

## **DeepSeek**

![__](https://upload.wikimedia.org/wikipedia/commons/thumb/e/ec/DeepSeek_logo.svg/1920px-DeepSeek_logo.svg.png)

**DeepSeek** (Chinese: 深度求索; pinyin: Shēndù Qiúsuǒ) is a Chinese artificial intelligence company that develops open-source large language models (LLMs). Based in Hangzhou, Zhejiang, it is owned and funded by Chinese hedge fund High-Flyer, whose co-founder, Liang Wenfeng, established the company in 2023 and serves as its CEO.

The DeepSeek-R1 model provides responses comparable to other contemporary large language models, such as OpenAI's GPT-4o and o1. It is trained at a significantly lower cost—stated at US$6 million compared to $100 million for OpenAI's GPT-4 in 2023—and approximately a tenth of the computing power used for Meta's comparable model, LLaMA 3.1. DeepSeek's AI models were developed amid United States sanctions on China and other countries for chips used to develop artificial intelligence, which were intended to restrict the ability of these countries to develop advanced AI systems. Lesser restrictions were later announced that would affect all but a few countries.

On 10 January 2025, DeepSeek released its first free chatbot app, based on the DeepSeek-R1 model, for iOS and Android; by 27 January, DeepSeek had surpassed ChatGPT as the most-downloaded free app on the iOS App Store in the United States,[10] causing Nvidia's share price to drop by 18%. DeepSeek's success against larger and more established rivals has been described as "upending AI"[10] and ushering in "a new era of AI brinkmanship". DeepSeek's compliance with Chinese government censorship policies and its data collection practices have also raised concerns over privacy and information control in the model, prompting regulatory scrutiny in multiple countries.

DeepSeek makes its generative artificial intelligence algorithms, models, and training details open-source, allowing its code to be freely available for use, modification, viewing, and designing documents for building purposes.However, reports indicate that the API version hosted in China applies content restrictions in accordance with local regulations, limiting responses on topics such as the Tiananmen Square massacre and Taiwan’s status. The company reportedly vigorously recruits young AI researchers from top Chinese universities, and hires from outside the computer science field to diversify its models' knowledge and abilities.

**Background**

In February 2016, High-Flyer was co-founded by AI enthusiast Liang Wenfeng, who had been trading since the 2007–2008 financial crisis while attending Zhejiang University. They began stock-trading with a deep learning model running on GPU on October 21, 2016. Prior to this, they used CPU-based models, mainly linear models. Most trading was done by AI by the end of 2017.

By 2019, he established High-Flyer as a hedge fund focused on developing and using AI trading algorithms. By 2021, High-Flyer exclusively used AI in trading, often using Nvidia chips. DeepSeek has made its generative artificial intelligence chatbot open source, meaning its code is freely available for use, modification, and viewing. This includes permission to access and use the source code, as well as design documents, for building purposes.

In 2021, while running High-Flyer, Liang began stockpiling Nvidia GPUs for an AI project.[20] According to 36Kr, Liang had built up a store of 10,000 Nvidia A100 GPUs, which are used to train AI, before the United States federal government imposed AI chip restrictions on China.

On 14 April 2023,[22] High-Flyer announced the start of an artificial general intelligence lab dedicated to research developing AI tools separate from High-Flyer's financial business. Incorporated on 17 July 2023, with High-Flyer as the investor and backer, the lab became its own company, DeepSeek. Venture capital firms were reluctant to provide funding, as they considered it unlikely that the venture would be able to generate an "exit" in a short period of time.

On May 16, 2023, the company Beijing DeepSeek Artificial Intelligence Basic Technology Research Co., Ltd. incorporated under the control of Hangzhou DeepSeek Artificial Intelligence Basic Technology Research Co., Ltd. As of May 2024, Liang Wenfeng held 84% of DeepSeek through two shell corporations.

After releasing DeepSeek-V2 in May 2024, which offered strong performance for a low price, DeepSeek became known as the catalyst for China's AI model price war. It was quickly dubbed the "Pinduoduo of AI", and other major tech giants such as ByteDance, Tencent, Baidu, and Alibaba began to cut the price of their AI models to compete with the company. Despite the low price charged by DeepSeek, it was profitable compared to its rivals that were losing money.

DeepSeek is focused on research and has no detailed plans for commercialization, which also allows its technology to avoid the most stringent provisions of China's AI regulations, such as requiring consumer-facing technology to comply with the government's controls on information.

DeepSeek's hiring preferences target technical abilities rather than work experience, resulting in most new hires being either recent university graduates or developers whose AI careers are less established. Likewise, the company recruits individuals without any computer science background to help its technology understand other topics and knowledge areas, including being able to generate poetry and perform well on the notoriously difficult Chinese college admissions exams (Gaokao).

**Training framework**

High-Flyer/DeepSeek has built at least two computing clusters, Fire-Flyer (萤火一号) and Fire-Flyer 2 (萤火二号). Fire-Flyer began construction in 2019 and finished in 2020, at a cost of 200 million yuan. It contained 1,100 GPUs interconnected at a rate of 200 Gbps. It was 'retired' after 1.5 years in operation. Fire-Flyer 2 began construction in 2021 with a budget of 1 billion yuan.[18] It was reported that in 2022, Fire-Flyer 2's capacity had been utilized at over 96%, totaling 56.74 million GPU hours. Of those GPU hours, 27% was used to support scientific computing outside the company.

Fire-Flyer 2 consisted of co-designed software and hardware architecture. On the hardware side, there are more GPUs with 200 Gbps interconnects. The cluster is divided into two "zones", and the platform supports cross-zone tasks. The network topology was two fat trees, chosen for its high bisection bandwidth. On the software side, there are

* **3FS (Fire-Flyer File System):** A distributed parallel file system. It was specifically designed for asynchronous random reads from a dataset, and uses Direct I/O and RDMA Read. In contrast to standard Buffered I/O, Direct I/O does not cache data. Caching is useless for this case, since each data read is random, and would not be reused.
* **hfreduce:** Library for asynchronous communication, originally designed to replace Nvidia Collective Communication Library (NCCL).[30] It was mainly used for allreduce, especially of gradients during backpropagation. It is asynchronously run on the CPU to avoid blocking kernels on the GPU.[28] It uses two-tree broadcast like NCCL.
* **hfai.nn:** Software library of commonly used operators in neural network training, similar to torch.nn in PyTorch.
* **HaiScale Distributed Data Parallel (DDP):** Parallel training library that implements various forms of parallelism in deep learning such as Data Parallelism (DP), Pipeline Parallelism (PP), Tensor Parallelism (TP), Experts Parallelism (EP), Fully Sharded Data Parallel (FSDP) and Zero Redundancy Optimizer (ZeRO). It is similar to PyTorch DDP, which uses NCCL on the backend.
* **HAI Platform:** Various applications such as task scheduling, fault handling, and disaster recovery.
During 2022, Fire-Flyer 2 had 5000 PCIe A100 GPUs in 625 nodes, each containing 8 GPUs. At the time, they chose to exclusively use PCIe instead of DGX version of A100, since at the time the models they trained could fit within a single 40 GB GPU VRAM, so there was no need for the higher bandwidth of DGX (i.e. they required only data parallelism but not model parallelism).[30] Later, they also incorporated NVLinks and NCCL, to train larger models that required model parallelism.